crates/driver/src/domain/competition/solution/scoring.rs

```diff
@@ -209,6 +173,54 @@ impl Trade {
         })
     }
 
+    /// Protocol fee as a cut of surplus, denominated in SURPLUS token
+    fn surplus_fee(&self, factor: f64) -> Result<eth::Asset, Error> {
+        let surplus = self.surplus().ok_or(Error::Surplus(self.sell, self.buy))?;
+        let fee = surplus
+            .amount
+            .apply_factor(factor / (1.0 - factor))
+            .ok_or(Error::Factor(surplus.amount, factor))?;
+
+        Ok(eth::Asset {
+            token: surplus.token,
+            amount: fee,
+        })
+    }
+
+    /// Protocol fee as a cut of the trade volume, denominated in SURPLUS token
+    fn volume_fee(&self, factor: f64) -> Result<eth::Asset, Error> {
+        let executed_in_surplus_token: eth::TokenAmount = match self.side {
+            Side::Sell => self
+                .executed
+                .0
+                .checked_mul(self.custom_price.sell)
+                .ok_or(Math::Overflow)?
+                .checked_div(self.custom_price.buy)
+                .ok_or(Math::DivisionByZero)?,
+            Side::Buy => self
+                .executed
+                .0
+                .checked_mul(self.custom_price.buy)
+                .ok_or(Math::Overflow)?
+                .checked_div(self.custom_price.sell)
+                .ok_or(Math::DivisionByZero)?,
+        }
+        .into();
+        let factor = match self.side {
+            Side::Sell => factor / (1.0 - factor),
+            Side::Buy => factor / (1.0 + factor),
+        };
+
+        Ok(eth::Asset {
+            token: self.surplus_token(),
+            amount: {
+                executed_in_surplus_token
+                    .apply_factor(factor)
+                    .ok_or(Error::Factor(executed_in_surplus_token, factor))?
```

Comment:
I still think this is essentially a Math::overflow error. What is the value of having `executed_in_surplus_token` and `factor` in the error type?

Reply:
It captures the `factor = 1.0` case. This case shouldn't be allowed because it's theoretically impossible to recalculate the protocol fee in those cases. For example, surplus fee is recreated with:

`// If the surplus after all fees is X, then the original surplus before protocol fee is X / (1 - factor)`

So, if the factor is 1.0, it means that the whole surplus is taken as fee, so order is executed at it's limit price. The formula would return:
X / (1 - factor) = 0 / inf = ?

Even if autopilot makes sure it never sends 1.0 factor, I think it's worth while to have this error handling. Although we might handle it some other way if you find this unclear.

Reply:
If factor is 1 this is basically `Math::DivisionByZero` (in both cases it's a Math error). Well formed-ness of the factor should be checked when we create the fee policy, so I don't expect this case to happen in practice (ie I don't foresee us to do custom error handling warranting a separate type).

Reply:
Removed Factor variant. Captured work around adding validation check: https://github.com/cowprotocol/services/issues/2501